src/Snap/CORS.hs

```diff
@@ -0,0 +1,280 @@
+--
+-- > applyCORS defaultOptions $ method POST $ myHandler
+applyCORS :: Snap.MonadSnap m => CORSOptions m -> m () -> m ()
+applyCORS options m =
+  (join . fmap decodeOrigin <$> getHeader "Origin") >>= maybe m corsRequestFrom
+
+ where
+
+  corsRequestFrom origin = do
+    originList <- corsAllowOrigin options
+    if origin `inOriginList` originList
+       then Snap.method Snap.OPTIONS (preflightRequestFrom origin)
+              <|> handleRequestFrom origin
+       else m
+
+  preflightRequestFrom origin = do
+    maybeMethod <- fmap (parseMethod . Char8.unpack) <$>
+                     getHeader "Access-Control-Request-Method"
+
+    case maybeMethod of
+      Nothing -> m
+
+      Just method -> do
+        allowedMethods <- corsAllowedMethods options
+
+        if method `HashSet.member` allowedMethods
+          then do
+            maybeHeaders <-
+              fromMaybe (Just HashSet.empty) . fmap splitHeaders
+                <$> getHeader "Access-Control-Request-Headers"
+
+            case maybeHeaders of
+              Nothing -> m
+              Just headers -> do
+                allowedHeaders <- corsAllowedHeaders options headers
+
+                if not $ HashSet.null $ headers `HashSet.difference` allowedHeaders
+                   then m
+                   else do
+                     addAccessControlAllowOrigin origin
+                     addAccessControlAllowCredentials
+
+                     commaSepHeader
+                       "Access-Control-Allow-Headers"
+                       Char8.pack (HashSet.toList allowedHeaders)
+
+                     commaSepHeader
+                       "Access-Control-Allow-Methods"
+                       (Char8.pack . show) (HashSet.toList allowedMethods)
+
+          else m
+
+  handleRequestFrom origin = do
+    addAccessControlAllowOrigin origin
+    addAccessControlAllowCredentials
+
+    exposeHeaders <- corsExposeHeaders options
+    when (not $ HashSet.null exposeHeaders) $
+      commaSepHeader
+        "Access-Control-Expose-Headers"
+        CI.original (HashSet.toList exposeHeaders)
+
+    m
+
+  addAccessControlAllowOrigin origin =
+    addHeader "Access-Control-Allow-Origin"
+              (encodeUtf8 $ Text.pack $ show origin)
+
+  addAccessControlAllowCredentials = do
+    allowCredentials <- corsAllowCredentials options
+    when (allowCredentials) $
+      addHeader "Access-Control-Allow-Credentials" "true"
+
+  decodeOrigin = fmap simplifyURI . parseURI . Text.unpack . decodeUtf8
+
+  addHeader k v = Snap.modifyResponse (Snap.addHeader k v)
+
+  commaSepHeader k f vs =
+    case vs of
+      [] -> return ()
+      _  -> addHeader k $ Char8.intercalate ", " (map f vs)
+
+  getHeader = Snap.getsRequest . Snap.getHeader
+
+  splitHeaders =
+    let spaces = Attoparsec.many' Attoparsec.space
+        headerC = Attoparsec.satisfy (not . (`elem`( " ," :: String)))
+        headerName = Attoparsec.many' headerC
```

Comment:
Sigh ... trying to do right by your review here. But: `pToken` is busted, unless I'm missing something. It will accept `,` as part of a token: 
```
λ> fullyParse "hello,goodbye" pToken
Right "hello,goodbye"
```